src/picom.c

```diff
@@ -1382,16 +1518,107 @@ static void unredirect(session_t *ps) {
 	free(ps->damage_ring);
 	ps->damage_ring = ps->damage = NULL;
 
+	if (ps->present_event_id) {
+		xcb_present_select_input(ps->c, ps->present_event_id,
+		                         session_get_target_window(ps), 0);
+		ps->present_event_id = XCB_NONE;
+		xcb_unregister_for_special_event(ps->c, ps->present_event);
+		ps->present_event = NULL;
+	}
+
 	// Must call XSync() here
 	x_sync(ps->c);
 
 	ps->redirected = false;
 	log_debug("Screen unredirected.");
 }
 
+static void
+handle_present_complete_notify(session_t *ps, xcb_present_complete_notify_event_t *cne) {
+	if (cne->kind != XCB_PRESENT_COMPLETE_KIND_NOTIFY_MSC) {
+		return;
+	}
+
+	bool event_is_invalid = false;
+	if (ps->frame_pacing) {
+		auto next_msc = cne->msc + 1;
+		if (cne->msc <= ps->last_msc || cne->ust == 0) {
+			// X sometimes sends duplicate/bogus MSC events, don't
+			// use the msc value. Also ignore these events.
+			//
+			// See:
+			// https://gitlab.freedesktop.org/xorg/xserver/-/issues/1418
+			next_msc = ps->last_msc + 1;
+			event_is_invalid = true;
+		}
+		auto cookie = xcb_present_notify_msc(ps->c, session_get_target_window(ps),
+		                                     0, next_msc, 0, 0);
+		set_cant_fail_cookie(ps, cookie);
+	}
+	if (event_is_invalid) {
+		return;
+	}
```

Comment:
it could be a single `if`.

Reply:
you mean like move the second if inside the first one?

Reply:
yeah, both variable definition and check if it's true could be in the `if (ps->frame_pacing)`.